README.md
<h1>KinashServer</h1>
KinashServer is simple node.JS HTTP server with some basic config
<h2>Launch</h2>
<p>To launch, use these commands: 
<br>
<br>
npm install kinashserver
<br>
"C:\users\%username%\node_modules\kinashserver\kinashserver.cmd"</p>
<h2>Download</h2>
<a href="https://github.com/andriy332/KinashServer/releases/">Click here to download</a>
<h2>Bugs</h2>
<a href="https://github.com/andriy332/KinashServer/issues">Click here to report a bug</a>
<h3>Change logs</h3
<p>Open the changelog.txt file</p>


changelog.txt
1.8:

Added support for almost all mime types
Renamed changelogs.txt to changelog.txt
Renamed kinashserver.bat to kinashserver.cmd
Fixed kinashserver.bat
Added authentication_enabled
Removed useless check & Now KinashServer can handle long urls
Removed useless return & typo fixes
Deleted default-config.json
Tabulation in package.json
res.setHeader() is now sendHeader()
Better code
Returnerror() optimization
Better error pages system
Remove accessDenied()
Remove useless comments
Fixed a bug that errors shows as warnings in logs

1.6:
Added status(code)
Added sendHeader(header, value)
Added info(content)
Added warning(content)
Added error(content)
Added accessDenied(reason) 
Added request method check 
Added req.url.length > 9999 check
Fixed a bug with returnError()
Fixed console stylistics
Rewritten returnError(err, message, statusText)
Updated process.on('uncaughtException'...)
Updated logging
Updated authentication
Updated config.json
readfile() is now readFile()
Used `returnError(503, 'No index page')
Instand of `returnError(500, null, null)`
Removed unused check
Removed max_url_length_tip_do_not_change from config
Removed error500noauthpage from config
Removed errornoindexpage from config
Removed standardjs from dependencies

1.5.1:
Fixed some files

1.5:
Removed "blacklistedurls" functionality
Fixed spelling error in config.json
Created docs (available at https://andriy332.github.io/KinashServer/)
Added some checks
Added default-config.json
Updated debug.cmd
Fixed code stylistics errors

1.4.1:
Deleted install-all-libs.bat
Fixed debug.cmd

1.4:

Fixed tabulation
Added Disallow crawlers functionality
Moved read-file code to readfile()
Used 
`throw new KinashServerError("A unknown error happend in user request! Please report this");`
instand of 
`console.error('\x1b[31m [ERROR] An error handling this user request')
console.error('\x1b[31m [ERROR] try { fail (148 line)')`
Updated error pages
Added package.json
Updated index.html


app.js
const http = require('http')
const fs = require('fs')
const mime = require('mime')
const rateLimit = require("http-ratelimit")
const config = require('./configs/config.json')
const server_version = '1.8.2'
const log = new console.Console(fs.createWriteStream('./logs/requests-log.txt'))
const errorlog = new console.Console(fs.createWriteStream('./logs/errors-log.txt'))


const server = http.createServer((req, res) => {

  function status(code) { res.statusCode = code }
  function sendHeader(header, value) { res.setHeader(header, value) }
  function writeContent(content) { res.write(content) }
  function endResponse(content) { res.end(content) }
  function info(content) { console.log('\x1b[0m\x1b[32m INFO >> ' + content); log.log('INFO >> ' + content) }
  function warning(content) { console.log('\x1b[0m\x1b[33m WARN >> ' + content); log.log('WARN >> ' + content); errorlog.log('WARN >> ' + content) }
  function error(content) { console.log('\x1b[0m\x1b[31m ERROR >> ' + content); log.log('ERROR >> ' + content); errorlog.log('ERROR >> ' + content) }

  function returnError(err, message, statusText) {
    status(err)
    sendHeader('Content-Type', 'text/html')
    if (err === 400 && message === null) { endResponse(config.error400page); return }
    if (err === 401 && message === null) { endResponse(config.error401page); return }
    if (err === 403 && message === null) { endResponse(config.error403page); return }
    if (err === 404 && message === null) { endResponse(config.error404page); return }
    if (err === 405 && message === null) { endResponse(config.error405page); return }
    if (err === 414 && message === null) { endResponse(config.error414page); return }
    if (err === 429 && message === null) { endResponse(config.error429page); return }
    if (err === 431 && message === null) { endResponse(config.error431page); return }
    if (err === 500 && message === null) { endResponse(config.error500page); }
    else { endResponse(`<html lang="en"><head><title>${err} ${statusText}</title></head><body><center><h1>${err} ${statusText}</h1><p>${message}</p></center></body></html>`) }
  }


  function readFile() {
    let read = './public_html' + req.url.split('?')[0]
    fs.readFile(read, 'utf8', (err, data) => {
      if (err) {
        returnError(404, null, null)
        return;
      }
      if (mime.getType(read).includes("image") == true || mime.getType(read).includes("audio") == true || mime.getType(read).includes("video") == true || mime.getType(read).includes("font") == true || mime.getType(read).includes("application") == true) {
        sendHeader('Content-type', mime.getType(req.url))
        let fileStream = fs.createReadStream(__dirname + '/./public_html' + req.url); //NOSONAR
        fileStream.pipe(res);
        return;
      }
      sendHeader('Content-type', mime.getType(read))
      endResponse(data)
    })
  };

  rateLimit.inboundRequest(req)

  process.on('uncaughtException', function (err) {
    returnError(500, null, null)
    error('Error handling this user request!')
    error('Please report this error to our github: https://github.com/andriy332/KinashServer/')
    error(err.stack)
    warning('Do not forget to send full server log')
    warning('And do not forget to send this details: ')
    warning('Server version: ' + server_version)
    warning('Request url: ' + req.url)
    warning('Request method: ' + req.method)
    warning('authentication_realm: ' + config.authentication_realm)
    warning('authentication_file: ' + config.authentication_file)
    warning('max_url_length: ' + config.max_url_length)
    warning('disallowcrawlers: ' + config.disallowcrawlers)
    warning('enablebasicsecuritychecks: ' + config.enablebasicsecuritychecks)
  })

  info(req.socket.remoteAddress + ' ' + req.method + ' ' + req.url + ' ' + Date())

  if (rateLimit.isRateLimited(req, config.ratelimit_maximumrequests) == true) {
    returnError(429, null, null)
    return;
  }

  else if (req.url === '/') {
    fs.open('./public_html/index.html', 'r', function (err, fileToRead) {
      if (!err) {
        fs.readFile(fileToRead, { encoding: 'utf-8' }, function (err, data) {
          if (!err) {
            sendHeader('Content-type', 'text/html')
            endResponse(data)
          } else {
            returnError(404, null, null)
          }
        })
      } else {
        returnError(500, null, null)
        error('The index.html file is missing')
      }
    })
  }

  else if (req.url.length > 10000) {
    returnError(431, null, null)
  } else if (req.url.includes("%") || req.url.includes("<") || req.url.includes(">") || req.url.includes("..")) {
    if (config.enablebasicsecuritychecks == true) {
      returnError(400, null, null)
      warning(req.socket.remoteAddress + ' tried to use exploit')
    }
    else { readFile() }
  } else if (req.url === config.authentication_url) {
    if (config.authentication_enabled == true) {
      const auth = { login: config.authentication_username, password: config.authentication_password }
      const b64auth = (req.headers.authorization || '').split(' ')[1] || ''
      const [login, password] = Buffer.from(b64auth, 'base64').toString().split(':')
      if (login && password && login === auth.login && password === auth.password) {
        fs.readFile(config.authentication_file, 'utf8', function (err, data) {
          if (err) {
            returnError(500, null, null)
            warning('User ' + req.socket.remoteAddress + ' passed the authentication')
            error('The authentication file is missing')
          }
          sendHeader('Content-Type', 'text/html')
          endResponse(data)
          warning('User ' + req.socket.remoteAddress + ' passed the authentication')
        })
        return
      }
      sendHeader('WWW-Authenticate', 'Basic realm="' + config.authentication_realm + '"')
      returnError(401, null, null)
      warning('User ' + req.socket.remoteAddress + ' is tried to login (or failed the authentication)')
    } else { readFile() }
  } else if (req.url.length > config.max_url_length) {
    returnError(414, null, null)
  } else if (req.url.includes('/login.html') == true) {
    returnError(403, null, null)
  } else if (req.url === '/robots.txt') {
    if (config.disallowcrawlers = true) {
      sendHeader('Content-type', 'text/plain')
      writeContent('User-agent: *')
      endResponse('Disallow: /')
    } else {
      readFile()
    }
  } else { readFile() }
})

server.listen(config.port, config.host, () => {
  //info() doesn't work here, so use console.log()
  console.log('\x1b[0m\x1b[32m INFO >> Loading server')
  rateLimit.init(config.ratelimit_time, true);
  if (config.port === "80") {
    console.log('\x1b[0m\x1b[32m INFO >> Server started at http://' + config.host + '/')
  }
  else {
    console.log('\x1b[0m\x1b[32m INFO >> Server started at http://' + config.host + ':' + config.port + '/')
  }
})


KinashServer_WithOutInstall/app1.js
const http = require("http");
const server_version = "1.3"
http.createServer(function(request, response){
    console.info('\x1b[32m User: ' + request.method + '  ' + request.url)
    console.info('\x1b[32m Server: 200 OK - ' + 'CustomRequestHeaders: Server: KinashServer')
    response.setHeader("Content-Type", "text/html; charset=utf-8;");

    if(request.url === "/" || request.url === "/home" ){
        response.setHeader("Server", "KinashServer");
        response.statusCode = 200;
        response.write('Index')
    }
    else if(request.url === "/old"){
 	console.info('\x1b[33m User: ' + request.method + '  ' + request.url)
  	console.info('\x1b[33m Server: 302 TempRedirect - ' + 'CustomRequestHeaders: Server: KinashServer')
        response.statusCode = 302;
        response.setHeader("Server", "KinashServer");
        response.setHeader("Location", "/new");
    }
    else if(request.url == "/new"){
        response.setHeader("Server", "KinashServer");
        response.write("<p>You are redirected from /old to /new</p>");
    }
    else if(request.url == "/Forbidden"){
 	console.info('\x1b[31m User: ' + request.method + '  ' + request.url)
  	console.info('\x1b[31m Server: 403 Forbidden - ' + 'CustomRequestHeaders: Server: KinashServer')
        response.setHeader("Server", "KinashServer");
        response.statusCode = 403;
	response.write("<center>");
	response.write("<h3>403 Forbidden </h3>");
	response.write("<p>KinashServer/" + server_version);
	response.write("</center>");
    }
    else{
 	console.info('\x1b[31m User: ' + request.method + '  ' + request.url)
  	console.info('\x1b[31m Server: 404 Not Found - ' + 'CustomRequestHeaders: Server: KinashServer')
    	response.statusCode = 404;
	response.write("<center>");
	response.write("<h3>404 Not Found </h3>");
	response.write("<p>KinashServer/" + server_version);
	response.write("</center>");
    }
    response.end();
}).listen(80);


src/server.js
const http = require('http')
const fs = require('fs')
const folder = '../public_html/'
const config = require('../configs/config.json')
const log = new console.Console(fs.createWriteStream('./logs/requests-log.txt'))
const errorlog = new console.Console(fs.createWriteStream('./logs/errors-log.txt'))
const errorslog = new console.Console(fs.createWriteStream('./logs/errors-log.txt'))
const mime = require('mime')
if (!fs.existsSync(folder)) {
  fs.mkdirSync(folder, { recursive: true })
}

const server = http.createServer((req, res) => {
  function returnError (err, message) {
    res.writeHead(err, { 'Content-Type': 'text/html' })
    if (err == '400') { res.end(config.error400page); return }
    if (err == '401') { res.end(config.error401page); return }
    if (err == '403') { res.end(config.error403page); return }
    if (err == '404') { res.end(config.error404page); return }
    if (err == '414') { res.end(config.error414page); return }
    if (err == '500') { res.end(config.error500page); return }
    res.write('<h3>Error </h3>')
    res.write(`<p>Error code: ${err}</p>`)
    res.end(`<u>${message}</u>`)
  }

  function readfile () {
    try {
      fs.readFile('./public_html' + req.url, 'utf8', (err, data) => {
        if (err) {
          returnError(404, null)
          return
        }
        res.setHeader('Content-type', mime.getType(req.url))
        res.end(data)
      })
    } catch (err) {
      returnError(500, null)
      throw new Error('A unknown error happend in user request! Please report this to our github')
    }
  };
  process.on('uncaughtException', function (err) {
    returnError(500, null)
    console.error('\x1b[31m [ERROR] An error handling in user request')
    console.warn('\x1b[33m [WARN] Please report this bug to our github')
    console.warn('\x1b[33m [WARN] ERROR:')
    console.warn('\x1b[33m' + err + '\x1b[0m')
    errorlog.error('[ERROR] An error handling in user request')
    errorlog.warn('[WARN] Please report this bug to our github')
    errorlog.warn('[WARN] ERROR:')
    errorlog.warn(err)
  })
  log.log('[INFO] ' + '[' + req.socket.remoteAddress + '] ' + Date() + ' ' + req.method + ' ' + req.url)
  console.log('[INFO] ' + '\x1b[0m\x1b[32m [' + req.socket.remoteAddress + '] ' + Date() + ' ' + req.method + ' ' + req.url)
  if (req.url == '/') {
    fs.open('./public_html/index.html', 'r', function (err, fileToRead) {
      if (!err) {
        fs.readFile(fileToRead, { encoding: 'utf-8' }, function (err, data) {
          if (!err) {
 	    res.writeHead(200, { 'Content-Type': 'text/html' })
            res.write(data)
            res.end()
          } else {
            returnError(404, null)
          }
        })
      } else {
	 returnError(503, "No index page")
      }
    })
  } else if (req.url == config.authentication_url) {
    const auth = { login: config.authentication_username, password: config.authentication_password }
    const b64auth = (req.headers.authorization || '').split(' ')[1] || ''
    const [login, password] = Buffer.from(b64auth, 'base64').toString().split(':')
    if (login && password && login === auth.login && password === auth.password) {
      fs.readFile(config.authentication_file, 'utf8', function (err, data) {
        if (err) {
	  returnError(503, "No authentication file found.")
	  console.warn('\x1b[33m[WARN] User ' + req.socket.remoteAddress + ' passed the authentication with but the authentication file doesnt exists \x1b[0m\x1b[32m')
        }
	 res.writeHead(200, { 'Content-Type': 'text/html' })
	 res.end(data)
	 console.warn('\x1b[33m[WARN] User ' + req.socket.remoteAddress + ' passed the authentication \x1b[0m\x1b[32m ')
      })
      return
    }
    res.setHeader('WWW-Authenticate', 'Basic realm="' + config.authentication_realm + '"')
    returnError(401, null)
    console.warn('\x1b[33m[WARN] User ' + req.socket.remoteAddress + ' is tried to login (or failed the authentication)')
  } else if (req.url.includes === '/%') {
    returnError(400, null)
  } else if (req.url.length > config.max_url_length) {
    returnError(414, null)
  } else if (req.url === config.blacklistedurls) {
    returnError(403, null)
  } else if (req.url === '/login.html') {
    returnError(404, null)
  } else if (req.url === '/robots.txt') {
    if (config.disallowcrawlers == 'true') {
      res.writeHead(200, { 'Content-Type': 'text/plain' })
      res.write('User-agent: *')
      res.write('Disallow: /')
      res.end('Disallow: *')
    } else {
      readfile()
    }
  } else if (req.url === '/login.html/') {
    returnError(403, null)
  } else {
 	readfile()
  }
})
server.listen(config.port, config.host, () => {
  console.log('\x1b[0m\x1b[32m[INFO] Server started at http://' + config.host + ':' + config.port + '/')
})
